source/common/upstream/cluster_manager_impl.h

```diff
@@ -332,10 +338,115 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
     std::vector<PerPriority> per_priority_update_params_;
   };
 
+  /**
+   * An implementation of an on-demand CDS handle. It forwards the discovery request to the cluster
+   * manager that created the handle.
+   *
+   * It's a protected type, so unit tests can use it.
+   */
+  class OdCdsApiHandleImpl : public OdCdsApiHandle {
+  public:
+    static OdCdsApiHandlePtr create(ClusterManagerImpl& parent, OdCdsApiSharedPtr odcds) {
+      return std::make_unique<OdCdsApiHandleImpl>(parent, std::move(odcds));
+    }
+
+    OdCdsApiHandleImpl(ClusterManagerImpl& parent, OdCdsApiSharedPtr odcds)
+        : parent_(parent), odcds_(std::move(odcds)) {
+      ASSERT(odcds_ != nullptr);
+    }
+
+    ClusterDiscoveryCallbackHandlePtr
+    requestOnDemandClusterDiscovery(const std::string& name,
+                                    ClusterDiscoveryCallbackSharedPtr callback,
+                                    std::chrono::milliseconds timeout) override {
+      return parent_.requestOnDemandClusterDiscovery(odcds_, name, std::move(callback), timeout);
+    }
+
+  private:
+    ClusterManagerImpl& parent_;
+    OdCdsApiSharedPtr odcds_;
+  };
+
   virtual void postThreadLocalClusterUpdate(ClusterManagerCluster& cm_cluster,
                                             ThreadLocalClusterUpdateParams&& params);
 
+  /**
+   * Notifies cluster discovery managers in each worker thread that
+   * the discovery process for the cluster with a passed name has
+   * timed out.
+   */
+  void notifyExpiredDiscovery(const std::string& name);
+
 private:
+  struct ThreadLocalClusterManagerImpl;
+
+  /** A thread-local on-demand cluster discovery manager. It takes care of invoking the discovery
+   *  callbacks in the event of a finished discovery. It does it by installing a cluster lifecycle
+   *  callback that invokes the discovery callbacks when a matching cluster just got added.
+   */
+  class ClusterDiscoveryManager : Logger::Loggable<Logger::Id::upstream> {
+  public:
+    ClusterDiscoveryManager(ThreadLocalClusterManagerImpl& parent);
+
+    /**
+     * Invoke the callbacks for the given cluster name. The discovery status is passed to the
+     * callbacks. After invoking the callbacks, they are dropped from the manager.
+     */
+    void processClusterName(const std::string& name, ClusterDiscoveryStatus cluster_status);
+
+    /**
+     * A pair containing a discovery handle and information whether a discovery for a given cluster
+     * was already requested in this thread.
+     */
+    struct Pair {
```

Comment:
Any reason not to use `std::pair`?

Reply:
Not a fan of `std::pair` with its opaque `first` and `second` members, really. :)